Detectors/TPC/workflow/src/RawToDigitsSpec.cxx

```diff
@@ -0,0 +1,209 @@
+        const auto dataWrapperID = ((subSpecification >> 8) & 0xFF) > 0;
+        const auto globalLinkID = linkID + dataWrapperID * 12;
+        const auto sector = cruID / 10;
+        processAttributes->activeSectors |= (0x1 << sector);
+
+        auto& reader = processAttributes->rawReader.getReaders()[0];
+        reader->forceCRU(cruID);
+        reader->setLink(globalLinkID);
+
+        //LOG(INFO) << dh->dataOrigin.as<std::string>() << "/" << dh->dataDescription.as<std::string>() << "/"
+        //<< dh->subSpecification << " payload size " << dh->payloadSize;
+        //LOG(INFO) << "CRU: " << cruID << " -- linkID: " << linkID << " -- dataWrapperID: " << dataWrapperID << " -- globalLinkID: " << globalLinkID;
+
+        // there is a bug in InpuRecord::get for vectors of simple types, not catched in
+        // DataAllocator unit test
+        //auto data = inputs.get<std::vector<char>>(input.spec->binding.c_str());
+        //LOG(INFO) << "data size " << data.size();
+        //printHeader();
+        try {
+          o2::framework::RawParser parser(input.payload, dh->payloadSize);
+
+          rawreader::ADCRawData rawData;
+          rawreader::GBTFrame gFrame;
+
+          for (auto it = parser.begin(), end = parser.end(); it != end; ++it) {
+            auto* rdhPtr = it.get_if<o2::header::RAWDataHeaderV4>();
+            if (!rdhPtr) {
+              break;
+            }
+            const auto& rdh = *rdhPtr;
+            //printRDH(rdh);
+            // ugly event handling ...
+            const auto hbOrbit = rdh.heartbeatOrbit;
+            const auto lastOrbit = processAttributes->lastOrbit;
+
+            if ((lastOrbit > 0) && (hbOrbit > (lastOrbit + 3))) {
+              auto& digitDump = processAttributes->digitDump;
+              digitDump.incrementNEvents();
+              LOG(INFO) << fmt::format("Number of processed events: {} ({})", digitDump.getNumberOfProcessedEvents(), processAttributes->maxEvents);
+              digitDump.sortDigits();
+              // add publish here
+              for (int isector = 0; isector < Sector::MAXSECTOR; ++isector) {
+                if (processAttributes->activeSectors & (0x1 << isector)) {
+                  snapshotDigits(digitDump.getDigits(isector), isector);
+                }
+              }
+              digitDump.clearDigits();
+
+              processAttributes->activeSectors = 0;
+              if (digitDump.getNumberOfProcessedEvents() >= processAttributes->maxEvents) {
+                LOG(INFO) << fmt::format("Maximm number of events reached ({}), no more processing will be done", processAttributes->maxEvents);
+                processAttributes->quit = true;
+                pc.services().get<ControlService>().endOfStream();
+                pc.services().get<ControlService>().readyToQuit(QuitRequest::All);
+                //pc.services().get<ControlService>().readyToQuit(QuitRequest::Me);
```

Comment:
Using `QuitRequest::All` will terminate all devices and some of them might not have processed the data yet, should be `QuitRequest::Me``